### DISCUSSION TITLE
Merge `filter` and `where`

### DISCUSSION THREAD

**Discussion by kaathewisegit** [nushell/nushell#12944] [General]
Since I've been having a roll with those lately, here's another one: having both `filter` and `where` is unnecessary, because they are almost the same.

Currently, the only difference is reading the condition closure from a variable: `filter $cond`.  But I think that this behavior can be added to `where`, because `where $arg` is meaningless without a condition:

```Nushell
~> ls | where $var
# does nothing
~> ls | where size < 1Gb and $arg
# lone closure is caught as an error
Error: nu::parser::unsupported_operation

  × boolean operation is not supported between bool and closure.
   ╭─[entry #1:1:12]
 1 │ ls | where size < 1Gb and $arg
   ·            ─────┬──── ─┬─ ──┬─
   ·                 │      │    ╰── closure
   ·                 │      ╰── doesn't support these values
   ·                 ╰── bool
   ╰────
```

As such, `where` can be modified to work in exactly the same way `filter` does, without making any backwards incompatible changes.  As such, I'd suggest doing that and keeping only one command, naming it either `where` or `filter`.

**Comment by fdncred** [comment]
`filter` doesn't support row conditions aka `$it`, `where` does. We separated them on purpose, but I don't remember all the details of why we did this. I'm sure someone could find the PR where `filter` was added and maybe it has some details.

**Reply by kaathewisegit** [reply]
Found the PR: #7365

It says this replaced `where -b`, but I'm not sure what this means.

@kubouch, would You kindly clue in a newby in how `where` used to work, in regards to this PR?

**Reply by kubouch** [reply]
`where` is a parser keyword using a special row condition syntax. Previously, there was the `-b` flag that took a closure, like the current `filter`, but IIRC it didn't work correctly. It also made the `where` command syntax very confusing and complicated, hence the split of the `-b` functionality to the `filter` command.

**Reply by fdncred** [reply]
It looks like `where -b` allowed `where` to take a closure. That's currently not allowed in `where` but is allowed in `filter`. So, `where` takes a RowCondition and `filter` takes a `Closure` now.

**Reply by kaathewisegit** [reply]
Got it.  Actually, it's kind of weird that `where` supports closures.  It seems to lead to stuff like this:

```Nushell
~> fs list | where name starts-with p and {|row| $row.size < 1Gb}
Error: nu::parser::unsupported_operation

  × boolean operation is not supported between bool and closure.
   ╭─[entry #1:1:17]
 1 │ fs list | where name starts-with p and {|row| $row.size < 1Gb}
   ·                 ─────────┬──────── ─┬─ ───────────┬───────────
   ·                          │          │             ╰── closure
   ·                          │          ╰── doesn't support these values
   ·                          ╰── bool
   ╰────
```

**Reply by fdncred** [reply]
if you look at the `where` signature, it does not support closures, only row condition.
```rust
    fn signature(&self) -> nu_protocol::Signature {
        Signature::build("where")
            .input_output_types(vec![
                (
                    Type::List(Box::new(Type::Any)),
                    Type::List(Box::new(Type::Any)),
                ),
                (Type::table(), Type::table()),
                (Type::Range, Type::Any),
            ])
            .required(
                "row_condition",
                SyntaxShape::RowCondition,
                "Filter condition.",
            )
            .allow_variants_without_examples(true)
            .category(Category::Filters)
    }
```

**Reply by kaathewisegit** [reply]
@fdncred for some reason, lone closures count as row conditions:

```Nushell
~> fs list | where {|row| $row.size < 4Kb}
╭───┬────────────┬──────┬───────┬───────────────╮
│ # │    name    │ type │ size  │   modified    │
├───┼────────────┼──────┼───────┼───────────────┤
│ 0 │ .gitignore │ file │  58 B │ 11 months ago │
│ 1 │ .vale.ini  │ file │ 134 B │ 3 weeks ago   │
│ 2 │ toolkit.nu │ file │ 667 B │ 3 days ago    │
╰───┴────────────┴──────┴───────┴───────────────╯
```

**Reply by kaathewisegit** [reply]
Yeah, it accepts plain blocks and conditions.

```Rust
pub fn parse_row_condition(working_set: &mut StateWorkingSet, spans: &[Span]) -> Expression {
    ...

    let block_id = match expression.expr {
        Expr::Block(block_id) => block_id,
        Expr::Closure(block_id) => block_id,
```

**Reply by fdncred** [reply]
I was wondering if RowConditions has them built in or something. Just hadn't got around to looking. Nice investigation.

**Reply by kaathewisegit** [reply]
I wonder if it'd make sense to remove closures from where.  Currently it's a bit weird, since it doesn't compose with the regular shapes

**Reply by kubouch** [reply]
Now I remembered that `where` can't take a closure from a variable: `let cond =  {|x| $x.type == dir }; where $cond`. For that you need `filter`.
